# -*- mode: Snakemake -*-
# Sequence Alignment: BLAT

rule align:
  input:
    SEQ=RUN_DIR + "/process/consolidated/{sample}.{read}.consol.fasta",
    GENOME=ROOT_DIR + "/" + config["RefGenomePath"]
  output:
    RUN_DIR + "/process/aligned/{sample}.{read}.psl"
  params: 
    config["BLATparams"]
  shell:
    "blat {input.GENOME} {input.SEQ} {output} {params}"

rule compress_align:
  input: RUN_DIR + "/process/aligned/{sample}.{type}.psl"
  output: RUN_DIR + "/process/aligned/{sample}.{type}.psl.gz"
  shell: "gzip {input}"

rule generate_2bit:
  input: "configs/" + RUN + ".config.yml"
  output: ROOT_DIR + "/" + config["RefGenomePath"]
  params: config["RefGenome"]
  run:
    R("""
    genome <- BSgenome.Hsapiens.UCSC.{params}::BSgenome.Hsapiens.UCSC.{params}
    BSgenome::export(genome, "{output}", format = "2bit")
    """)
